### PR TITLE
No-JS Search

### LINKED ISSUE #86: Make app work without JS
Currently, the complete application without JS except for the search form. We could make that work without JS as well though relatively easily.

### Search by search term

If we made the search form a `GET` form with `<form method="GET" action="/search">`, if the browser has JS switched off (or the app has not started so the `onsubmit` handler is not attached yet), that form would submit as e.g. `/search?search-term=munich` which we could simply redirect to `/search/munich` in the SSR server. `/search/munich` is already handled by the SSR server so searching by search term without JS enabled works 🚀 

### Search by coordinates

Without JS, we cannot get the users's coordinates in the browser. We can however, guess their rough coordinates based on their IP on the server side. So making searching by coordinates work without JS could simply work by making the current _"Show nearby locations"_ a link to e.g. `/search-by-coordinates` (that is handled by the app in the browsers via `onclick` once the app has started). When the SSR server sees a request to that path, it could simply get the user's coordinates based on their IP and redirect to e.g. `/search/23.3,45.5` which is already handled by the SSR server. Thus, searching for nearby locations would work without JS as well 🚀

### PR DESCRIPTION
**Pull request by jorgelainfiesta**
Closes #86

### CODE REVIEW

File: integration-tests/routes/index-test.js [hunk 1]
```diff
@@ -11,15 +11,15 @@ describe('the index route', function() {
...
+        expect($response('[data-test-search-input]').is(':disabled')).to.be.false;
       });
     });
 
-    it('disables the location search button', async function() {
+    it('enables the location search button', async function() {
       await visit('/', async (page, $response) => {
-        expect($response('[data-test-search-submit]').is(':disabled')).to.be.true;
+        expect($response('[data-test-search-submit]').is(':disabled')).to.be.false;
       });
```
Comment (marcoow) [inline integration-tests/routes/index-test.js]: I think these test cases can just be removed now.

File: integration-tests/routes/search-test.js [hunk 2]
```diff
@@ -19,10 +23,18 @@ describe('the search route', function() {
...
     });
 
-    it('renders a loader over the search form', async function() {
+    it('renders no loader over the search form', async function() {
       await visit('/search/Salzburg', async (page, $response) => {
         let element = $response('[data-test-form-laoder]');
 
+        expect(element.length).to.be.not.ok;
+      });
+    });
```
Comment (marcoow) [inline integration-tests/routes/search-test.js]: I think this test case could be removed now.

File: src/ui/components/Search/component.ts [hunk 3]
```diff
@@ -13,15 +13,15 @@ export default class Home extends Component {
...
+    coordinates: string;
     searchResults: string[];
     updateFogEffect: (index: number) => void;
     pullIndexedDB: () => void;
     router: Navigo;
   };
 
   @tracked
-  coordinates: number[] | null;
+  coordinates: string | null;
```
Comment (marcoow) [inline src/ui/components/Search/component.ts]: Hm, I don't understand why this change is necessary. Could you explain?

File: ssr-server.js [hunk 4]
```diff
@@ -79,7 +82,30 @@ async function preprender(req, res, next, data = {orbit: []}) {
...
+app.get('/search-by-coordinates', async function(req, res, next) {
+  try {
+    let ip = req.clientIp;
+    let geoData = geoip.lookup(ip);
+
+    if (geoData) {
+      let [ lat, lon ] = geoData.ll;
+      res.redirect(`/search/${lat},${lon}`);
+    } else {
+      res.redirect(`/search/0,0`);
```
Comment (marcoow) [inline ssr-server.js]: Can we use values that are out of range, e.g. `200,200` or so?

File: src/ui/components/Search/component.ts (hunk 3 above)
Comment (jorgelainfiesta) [reply src/ui/components/Search/component.ts]: It turned out that the actual value has been a plain string all this time. It was a mistake to have it as `number[]`.

File: ssr-server.js (hunk 4 above)
Comment (jorgelainfiesta) [reply ssr-server.js]: 0,0 is also an impossible value due the precision of coordinates; but you're not right, it's not semantically correct. I'll change it to a better value.

Comment (jorgelainfiesta) [reply ssr-server.js]: Correction: you're ~not~ right

File: src/ui/components/Search/component.ts (hunk 3 above)
Comment (marcoow) [reply src/ui/components/Search/component.ts]: > It turned out that the actual value has been a plain string all this time.

Hm, we should fix that instead then. I guess we'd only have to update the [routing code here](https://github.com/simplabs/breethe-client/blob/master/src/ui/components/Breethe/component.ts#L131).

Comment (jorgelainfiesta) [reply src/ui/components/Search/component.ts]: I'm looking into this but I'm not sure parsing and formatting is beneficial for us. The separated values are never used in the app. We'd have to parse it on the function you mentioned and then serialize them as a string again when we send them to the server.

If we decide to split it, I think we should keep the numbers as strings to avoid losing precision on the coordinates with JS' floats.

Comment (marcoow) [reply src/ui/components/Search/component.ts]: I think we could also treat the coordinates as strings but split them into an array of strings at least. That should be relatively easy by simply modifying the regexp actually I think?

File: src/ui/components/Search/component.ts [hunk 5]
```diff
@@ -37,14 +37,20 @@ export default class Home extends Component {
...
 
   @tracked('locations', 'searchTerm', 'coordinates')
   get showRecent(): boolean {
     return this.locations.length > 0 && !this.searchTerm && !this.coordinates;
   }
 
+  @tracked('coordinates')
+  get notFoundCoordinates(): boolean {
+    let { coordinates } = this;
+    return coordinates === '200,200';
```
Comment (marcoow) [inline src/ui/components/Search/component.ts]: I think we could use a union type for the coordinates so that we can express we either expect an array of strings or e.g. the number `-1` (which stands for location not found then). I think that would make it easier than using this magic value?

```typescript
coordinates: string[] | number
```

**Comment by jorgelainfiesta** [comment]
@marcoow is there any other comment about this PR?

File: integration-tests/nojs-test.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+const { expect } = require('chai');
+const visit = require('./helpers/visit');
+
+describe('the main flow without javascript', function() {
+  it('works', async function() {
+    await visit('/', { disableJavascript: false }, async (page) => {
```
Comment (marcoow) [inline integration-tests/nojs-test.js]: Should this not be `disableJavascript: true`?

File: integration-tests/routes/search-test.js [hunk 7]
```diff
@@ -207,4 +226,27 @@ describe('the search route', function() {
       });
     });
   });
+
+  describe('without javascript', function() {
+    it('renders search results', async function() {
+      await visit('/search/Salzburg', { disableJavascript: true }, async(page, $response) => {
+
```
Comment (marcoow) [inline integration-tests/routes/search-test.js]: extra newline 👀

File: integration-tests/routes/search-test.js [hunk 8]
```diff
@@ -207,4 +226,27 @@ describe('the search route', function() {
...
+        await page.waitForSelector('[data-test-search-result="Salzburg"]');
+        await page.click('[data-test-search-result="Salzburg"] a');
+        await page.waitForSelector('[data-test-location]');
+  
+        expect(page.url()).to.have.path('/location/2');
+      });
+    });
```
Comment (marcoow) [inline integration-tests/routes/search-test.js]: I think this doesn't actually test that search results are rendered but that navigating from search results to details page works? That is already covered by the [test above](https://github.com/simplabs/breethe-client/pull/92/files#diff-577670e6f179cf7aa04d01a8c75ee48eR5) so I think can just be removed.

File: src/ui/components/Breethe/component.ts [hunk 9]
```diff
@@ -128,7 +128,8 @@ export default class Breethe extends Component {
         params = params as SearchParams;
         this.locationId = null;
         this.searchTerm = params.searchTerm;
-        this.coordinates = params.coordinates;
+        let { coordinates } = params;
+        this.coordinates = (coordinates) ? coordinates.match(/[-+]?[0-9]*\.?[0-9]+/g) : null;
```
Comment (marcoow) [inline src/ui/components/Breethe/component.ts]: Is this really necessary? If coordinates wouldn't actually match [here](https://github.com/simplabs/breethe-client/pull/92/files#diff-a0d31ced42dfb7f0cc15c63975bb82e8R117), they should be `null` anyway. I'd probably capture `lan` and `lon` separately in the router and pass them as an array into `_setMode`. That should be much simpler?

File: src/ui/components/SearchForm/component.ts [hunk 10]
```diff
@@ -18,14 +17,9 @@ export default class SearchForm extends Component {
 
   submitSearch(event) {
     event.preventDefault();
-    let { search } = this;
     this.args.onSearchByTerm(this.search);
   }
 
-  searchByLocation(event) {
```
Comment (marcoow) [inline src/ui/components/SearchForm/component.ts]: Is this not necessary anymore?

File: src/ui/components/SearchForm/template.hbs [hunk 11]
```diff
@@ -1,6 +1,14 @@
 <div>
-  <form onsubmit={{action submitSearch}} class="search-form">
-    <input value={{search}} oninput={{action updateSearch}} placeholder="Enter location" disabled={{@isDisabled}} class="search-term" data-test-search-input>
+  <form action="/search" onsubmit={{action submitSearch}} class="search-form">
```
Comment (marcoow) [inline src/ui/components/SearchForm/template.hbs]: This should be `method="GET"` - we should never send any `POST` requests to the SSR server.

File: ssr-server.js [hunk 12]
```diff
@@ -79,7 +82,30 @@ async function preprender(req, res, next, data = {orbit: []}) {
...
+app.get('/search-by-coordinates', async function(req, res, next) {
+  try {
+    let ip = req.clientIp;
+    let geoData = geoip.lookup(ip);
+
+    if (geoData) {
+      let [ lat, lon ] = geoData.ll;
+      res.redirect(`/search/${lat},${lon}`);
+    } else {
+      res.redirect(`/search/200,200`);
```
Comment (marcoow) [inline ssr-server.js]: Can we use a special URL for this instead of these magic coordinates? E.g. we could use something like `/search/location-not-found` and handle that in the router in the client which would also allow us to simplify [this](https://github.com/simplabs/breethe-client/pull/92/files#diff-e8b5345cbd8d68683d57ab03613d1135R49) significantly.

File: integration-tests/routes/search-test.js [hunk 13]
```diff
@@ -84,6 +88,21 @@ describe('the search route', function() {
...
+
+    it('redirects /search-by-coordinates', async function() {
+      await visit('/search-by-coordinates', async (page, $response) => {
+        expect(page.url()).to.have.path('/search/200,200');
+
+        let error = $response('[data-test-coordinates-error]');
+        expect(error).to.be.ok;
+      });
+    });
```
Comment (marcoow) [inline integration-tests/routes/search-test.js]: Can we have a test for when this succeeds as well?

File: integration-tests/nojs-test.js (hunk 6 above)
Comment (jorgelainfiesta) [reply integration-tests/nojs-test.js]: Oh yeah, thanks. I must have missed it when I was experimenting with configs.

File: src/ui/components/SearchForm/component.ts (hunk 10 above)
Comment (jorgelainfiesta) [reply src/ui/components/SearchForm/component.ts]: Searching by location is its [own component](https://github.com/simplabs/breethe-client/blob/nojs-search/src/ui/components/SearchCoordinates/component.ts#L8) now.

File: ssr-server.js (hunk 12 above)
Comment (jorgelainfiesta) [reply ssr-server.js]: I'll give this approach a try.

File: src/ui/components/SearchForm/component.ts (hunk 10 above)
Comment (marcoow) [reply src/ui/components/SearchForm/component.ts]: Ah, I see - so this was previously unused anyway…

File: src/ui/components/Breethe/component.ts [hunk 14]
```diff
@@ -114,7 +118,11 @@ export default class Breethe extends Component {
     this.router
       .on('/', () => this._setMode(MODE_SEARCH))
       .on('/search', () => this._setMode(MODE_SEARCH))
-      .on(/search\/(\-?\d+(\.\d+)?,\s*\-?\d+(\.\d+)?)/, (coordinates) => this._setMode(MODE_SEARCH, { coordinates }))
+      .on(/search\/(\-?\d+(\.\d+)?,\s*\-?\d+(\.\d+)?)/, (coordinatesRaw: string) => {
+        let coordinates: string[] = coordinatesRaw.match(/[-+]?[0-9]*\.?[0-9]+/g);
+        this._setMode(MODE_SEARCH, { coordinates });
+      })
```
Comment (marcoow) [inline src/ui/components/Breethe/component.ts]: I think we could simply change the regexp for the route here?

```js
.on(/search\/((\-?\d+(\.\d+)?),\s*(\-?\d+(\.\d+)?))/, (_, lat, _, lon) => {
  //only have to convert lat and lon to numbers then here
}
```

Comment (jorgelainfiesta) [reply src/ui/components/Breethe/component.ts]: Trues! I follow this pattern now with some changes to the expression.

File: src/ui/components/Search/component.ts [hunk 15]
```diff
@@ -37,14 +38,19 @@ export default class Home extends Component {
...
 
   @tracked('locations', 'searchTerm', 'coordinates')
   get showRecent(): boolean {
     return this.locations.length > 0 && !this.searchTerm && !this.coordinates;
   }
 
+  @tracked('coordinates')
+  get notFoundCoordinates(): boolean {
+    return this.args.locationNotFound;
+  }
```
[the remark this answers is not in the document]
Comment (marcoow) [reply src/ui/components/Search/component.ts]: I think this can now be deleted?

[the remark this answers is not in the document]
Comment (jorgelainfiesta) [reply src/ui/components/Search/component.ts]: This is indeed gone but for some reason it's still showing up here. You can check the current computed properties directly on the file for this branch: https://github.com/simplabs/breethe-client/blob/nojs-search/src/ui/components/Search/component.ts#L39

[the remark this answers is not in the document]
Comment (marcoow) [reply src/ui/components/Search/component.ts]: The property is actually showing on the page you linked to above…
